### ISSUE TITLE
frost(sdpa): THD zero-host-read execute loads uninitialized KV capacity rows — NaN poisons P@V on the f16/fp8 SM100/SM120 rows

### ISSUE THREAD

**Issue by Aneureka** [NVIDIA/cudnn-frontend#624]
## Summary

The zero-host-read THD execute (#606 for SM100, #608 for SM120) binds K/V views at their
buffers' **capacity** — the packed token totals are device values, so the host can no longer
size the views to the live totals. The K/V TMA descriptors therefore span rows **between the
valid KV length and the storage end**: uninitialized memory the caller never wrote.

Those rows are neutralized on the S side (the shared `apply_mask_chunk` / the SM120 mask
overwrite masked columns with `-inf` via a **select**, which is NaN-safe), but nothing
protects the **V** side: the masked columns' P is `exp(-inf) = 0`, and in `P @ V` the IEEE
rule `0 * NaN = NaN` lets a NaN bit pattern in a garbage V row poison the whole accumulator —
every output row of the tile becomes NaN.

Before the zero-host-read rework this could not happen: the views (and descriptor extents)
were sized to the live totals read on the host, so rows past the packed total were outside
the descriptor and TMA **hardware-zero-filled** them.

The failure is **nondeterministic**: it needs NaN bit patterns in recycled device memory
(fresh CUDA allocations are usually zero pages, which is why the THD suites — whose harnesses
zero-fill their pads — stay green). f16 is especially exposed: ~3% of random bytes decode to
an f16 NaN, so a single 128-element garbage row almost surely contains one. In the wild it
surfaces as nondeterministic whole-sequence NaN on
`test_mhas_v2.py::test_sdpa_fp8_fwd_ragged_L0` (the mhas harness binds rounded-up
allocations).

## Affected

| Engine | Serves THD | Exposed | Evidence |
|---|---|---|---|
| `sdpa_fwd_prefill_sm120` (f16/bf16) | yes | **yes — reproduced** (RTX 5080 & RTX PRO 6000 SE) | probe below: NaN pad → `nan_frac=1.000` |
| `sdpa_fwd_prefill_sm100_d128` (f16/bf16) | yes | **yes — reproduced** (SM100) | same probe, same result |
| `sdpa_fwd_prefill_sm120_fp8` | yes | **yes — same mechanism** | fp8 variant of the probe reproduces it; also the mhas ragged NaN above |
| `sdpa_fwd_prefill_sm100_d128_fp8` | no (`thd=False`; ragged fp8 routes to the native backend) | n/a | mismatch probe: "graph uses THD / ragged, which this engine does not support" |
| `sdpa_fwd_prefill_sm100_d128_mxfp8` | (check) | untested | — |

## Repro

Minimal THD graph (b=1, h=1, D=128, lens q=[35] / kv=[190]) with the KV storage allocated at
a rounded-up capacity (192 tokens) and the 2 capacity-tail rows explicitly filled with NaN
(standing in for recycled memory):

- capacity pad = zeros → PASS (max|diff| = 0)
- capacity pad = NaN  → **FAIL: every output row NaN** (`nan_frac=1.000`), on both f16 rows

Observed:

```
sm100 f16 cap exact,       clean pad: max|diff|=0.0000 nan_frac=0.000 -> PASS
sm100 f16 cap padded(192), clean pad: max|diff|=0.0000 nan_frac=0.000 -> PASS
sm100 f16 cap padded(192), NaN pad:   max|diff|=nan    nan_frac=1.000 -> FAIL
sm120 f16: identical pattern
```

<details>
<summary>Probe script (run from <code>test/python</code> with FROST enabled; <code>python probe.py sm100|sm120</code>)</summary>

```python
import math, os, sys
sys.path.insert(0, os.path.join(os.getcwd(), "sdpa", "frost"))
import torch, cudnn
from cudnn.sdpa.fwd.engines import engine_name
from frost_test_utils import select_engine

arch = sys.argv[1] if len(sys.argv) > 1 else "sm100"
dev, D, h = "cuda", 128, 1

def run(label, q_cap, kv_cap, poison):
    lens_q, lens_kv = [35], [190]
    scale = 1.0 / math.sqrt(D)
    torch.manual_seed(0)
    q_f = [(torch.randn(1, h, n, D, device=dev) * 0.5).half() for n in lens_q]
    k_f = [(torch.randn(1, h, n, D, device=dev) * 0.5).half() for n in lens_kv]
    v_f = [(torch.randn(1, h, n, D, device=dev) * 0.5).half() for n in lens_kv]
    def pack(seqs, cap):
        st = torch.zeros(cap * h * D, dtype=torch.float16, device=dev)
        cu = [0]
        for s in seqs:
            n = s.shape[2]
            st[cu[-1] * h * D : (cu[-1] + n) * h * D] = s[0].permute(1, 0, 2).reshape(-1)
            cu.append(cu[-1] + n)
        return st, cu
    q_st, cu_q = pack(q_f, q_cap); k_st, cu_k = pack(k_f, kv_cap); v_st, _ = pack(v_f, kv_cap)
    if poison:
        for st in (k_st, v_st):
            st[sum(lens_kv) * h * D :] = float("nan")
    o_st = torch.full((q_cap * h * D,), float("nan"), dtype=torch.float16, device=dev)
    s_qd, s_kd = max(lens_q), max(lens_kv)
    ro_q = (torch.tensor(cu_q, dtype=torch.int64, device=dev) * h * D).view(2, 1, 1, 1)
    ro_k = (torch.tensor(cu_k, dtype=torch.int64, device=dev) * h * D).view(2, 1, 1, 1)
    g = cudnn.pygraph(io_data_type=cudnn.data_type.HALF, intermediate_data_type=cudnn.data_type.FLOAT, compute_data_type=cudnn.data_type.FLOAT)
    t4 = lambda shape, stride, dt: g.tensor(dim=shape, stride=stride, data_type=dt)
    tq = t4([1, h, s_qd, D], [s_qd * h * D, D, h * D, 1], cudnn.data_type.HALF)
    tk = t4([1, h, s_kd, D], [s_kd * h * D, D, h * D, 1], cudnn.data_type.HALF)
    tv = t4([1, h, s_kd, D], [s_kd * h * D, D, h * D, 1], cudnn.data_type.HALF)
    rq = g.tensor_like(ro_q); rk = g.tensor_like(ro_k); rv = g.tensor_like(ro_k); ro = g.tensor_like(ro_q)
    tq.set_ragged_offset(rq); tk.set_ragged_offset(rk); tv.set_ragged_offset(rv)
    sq_t = torch.tensor(lens_q, dtype=torch.int32, device=dev).view(1, 1, 1, 1)
    skv_t = torch.tensor(lens_kv, dtype=torch.int32, device=dev).view(1, 1, 1, 1)
    sq_h = g.tensor_like(sq_t); skv_h = g.tensor_like(skv_t)
    o, stats = g.sdpa(q=tq, k=tk, v=tv, attn_scale=scale, generate_stats=False,
                      use_padding_mask=True, seq_len_q=sq_h, seq_len_kv=skv_h)
    o.set_output(True).set_dim([1, h, s_qd, D]).set_stride([s_qd * h * D, D, h * D, 1]).set_data_type(cudnn.data_type.HALF)
    o.set_ragged_offset(ro)
    vp = {tq: q_st.view(-1, h, D), tk: k_st.view(-1, h, D), tv: v_st.view(-1, h, D), o: o_st.view(-1, h, D),
          rq: ro_q, rk: ro_k, rv: ro_k, ro: ro_q, sq_h: sq_t, skv_h: skv_t}
    g.validate(); g.build_operation_graph(); g.create_execution_plans([cudnn.heur_mode.A])
    select_engine(g, engine_name(128, arch=arch) if arch == "sm100" else engine_name(arch=arch))
    g.check_support(); g.build_plans()
    g.execute(vp, torch.empty(max(g.get_workspace_size(), 1), device=dev, dtype=torch.uint8))
    torch.cuda.synchronize()
    got = o_st[: 35 * h * D].view(35, h, D).float()
    ref = torch.softmax((q_f[0][0, 0].float() @ k_f[0][0, 0].float().T) * scale, dim=-1) @ v_f[0][0, 0].float()
    diff = (got[:, 0] - ref).abs().max().item()
    nanf = torch.isnan(got).float().mean().item()
    print(f"{label}: max|diff|={diff:.4f} nan_frac={nanf:.3f} -> {'FAIL' if (nanf > 0 or diff > 0.05) else 'PASS'}")

run(f"{arch} f16 cap exact, clean pad", 35, 190, False)
run(f"{arch} f16 cap padded(192), clean pad", 64, 192, False)
run(f"{arch} f16 cap padded(192), NaN pad", 64, 192, True)
```

</details>

## Suggested fixes

1. Per-kernel: zero the sV overhang rows in the masked steps before the PV MMA — only masked
   steps can overhang the valid KV extent, so the unmasked steady state is untouched.
2. Centralized alternative: the THD setup kernel already computes `cu_k[B]` device-side —
   it could zero the K/V rows in `[cu_k[B], capacity)` once per execute, fixing every
   consumer at one site (cost: bounded by the allocation rounding, typically < 1 tile).

Either way, a dirty-KV-pad THD test (poison the capacity tail with NaN before binding)
should join the suites — zero-filled harness pads are why this stayed invisible.

## References

- #606 (SM100 zero-host-read THD), #608 (SM120 port) — the capacity-view contract

**Comment by Aneureka** [comment]
This commit https://github.com/NVIDIA/cudnn-frontend/pull/595/changes/bdf5d12d371b7b01118af8982b22419884b81584 explicitly sanitizes the V tail in the SM120 SDPA FP8 kernel for THD cases to address the NaN issue. However, we still need to fix the issue in the other SDPA kernels and may find a better approach.

**Comment by vedaanta** [comment]
There is a **third fix already implemented in-tree**, and it is cheaper than either option in the issue: the FP8/MXFP8 rows solve this by *clamping the descriptor extent*, not by zeroing memory.

`build_thd_meta_o_kv_descs_kernel` (`thd_sm100.py`) copies the base K and V TensorMaps into `o_desc_words` slots `n_batch+1`/`n_batch+2` and patches `GLOBAL_DIM ord=2` to `cu_k[B]` — read from the metadata the same elected thread just wrote — then releases `GENERIC -> TENSORMAP`. The mainloop rebinds its load closures under `CFG.THD_VARLEN` (`prefill_d128_fp8_sm100.py:642`) so every load site stays branch-free. Tile-tail rows past the packed total become TMA-OOB and land as **exact zeros in SMEM without any memory access**.

Versus the two options listed:

- It touches **no caller memory**. Option 2 memsets `[cu_k[B], capacity)`, which both costs bandwidth and clobbers bytes the caller owns and may care about.
- It costs no per-execute traffic at all — the OOB rows are never fetched, rather than fetched-then-fixed.
- It needs no per-kernel masking change (option 1), so it cannot drift out of sync across kernel flavors.
- It is already validated: three FP8 kernels ship it today.

### Confirmed differential repro (SM100, bf16, cuDNN 9.26.0.33, FROST forced)

THD graph, `seq_lens = [200, 150, 47]` (total 397, deliberately not a multiple of `TILE_N`), K/V bound as `(CAP, H, D)` buffers whose `[397, CAP)` tail is NaN. Only the tail fill varies:

| binding | NaNs in O |
|---|---|
| `CAP = 640` (> total), NaN tail | **201,728 — 49.6% of O** |
| `CAP = 640`, zero tail | 0 |
| `CAP = 397` (== total, extent exact), NaN tail | **0** |

The third row is the control that isolates the mechanism: with the extent equal to the packed total the tail is unreachable and NaN there is harmless. This is exactly the state the clamp produces. Script: `probe_f16_tail.py` in #718.

### Scope — which tensors actually need it

The clamp is needed precisely where a tensor sits on the **contracted** side of a matmul:

- **V** — contracted in BMM2, `O[i,:] = sum_j P[i,j]*V[j,:]`. One NaN row *j* reaches *every* output row via `0*NaN`. Needs it.
- **K** — contracted in BMM1. A NaN row *j* poisons column *j*; the mask select normally neutralizes that, **except** on cc10.3 where the fused-LDTM row-max reduces S before the mask. Needs it.
- **Q** — the parallel dimension. A garbage Q row poisons only its own S/P/O row, and that row is never stored (dead-unit decode plus the per-sequence O extent). Not needed for NaN-safety — though clamping it from `cu_q[B]` anyway would retire the host-derived capacity entirely, since `tensormap_replace` overwrites the dim and so makes the tensor immune to both over-claim (this issue) and under-claim (#613).
- **O** — already stronger than clamped: per-sequence descriptors with per-sequence extents. **LSE** — predicated per row against the device-read length. Neither needs changing.

### Files

`prefill_d128_f16_sm100.py` and `prefill_d512_f16_sm100.py` call the O-only `build_thd_meta_o_descs_kernel`; they need `o_desc_slots = b + 3` (currently `b + 1`), the `_o_kv_descs` variant, and the two closure swaps. `prefill_f16_sm120.py` is on the meta-only `build_thd_meta_kernel`, so the K/V slots have to be introduced there rather than switched.

The one thing to measure is the `fence.proxy.tensormap::generic.acquire` landing on the K/V load path; `tma_tensormap_acquire` exists to hoist it to once per CTA (`acquire=False` per tile) instead of per load.

### Padded THD would make this much worse

Worth recording while this is open. In **packed** THD an interior tile tail steps into the *next sequence's real data* — finite, harmless — so only the final sequence's tail is exposed, one tile per launch. In **TE-style padded THD** (`cu_seqlens_padded != cu_seqlens`) every sequence's tail tile lands in *its own pad slot*, so the exposure becomes B tails instead of 1.

TE has both flavors: the identifier-token case documented in `DotProductAttention` (`[a a a 1 b b 2 2 c c c c 3]`, `cu_seqlens=[0,3,5,9]` vs `cu_seqlens_padded=[0,4,8,13]`) holds real embeddings in the pad slots, but the context-parallel case pads sequences to uniform `2*cp_size` chunks with no initialization guarantee. The THD lowering does not serve padded THD today — and per `engines.py` cannot decline it at plan time, since the offsets are runtime data — so this is latent rather than live, but it means the clamp stops being optional the moment padded THD is served.

---

Filed #718 for the f16-vs-FP8 gap before finding this issue; closing it as a duplicate and consolidating here.

**Comment by vedaanta** [comment]
Opened #740, which gives callers a way to sidestep this entirely: the forward SDPA node now accepts `max_total_seq_len_q/kv`, which `sdpa_backward` has taken since cuDNN 9.6 but forward never did. Declaring the packed total makes the THD extent exact, so the over-allocated tail is out of TMA reach.

Measured on the same repro from my comment above (`seq_lens=[200,150,47]`, `CAP=640`, NaN tail): **201,728 NaNs in O (49.6%) undeclared → 0 declared**, bit-identical to the zero-tail run.

Complementary rather than a replacement — a caller who declares nothing still gets the inferred bound, so porting the FP8 descriptor clamp to f16 remains the fix that closes this for everyone. #740 just means integrations that already hold the total (all of vLLM/SGLang/TE/Megatron/PyTorch/FlashInfer do — it is `q.shape[0]`) can opt out today.

**Comment by vedaanta** [comment]
**Fix implemented for SM100 f16/bf16 — PR #646.**

Following up on the earlier comment proposing the descriptor clamp as a third option (better than zeroing sV per-kernel or memsetting `[cu_k[B], capacity)` in the setup kernel): it is implemented and green.

`build_thd_meta_o_descs_kernel` now takes the K/V base TensorMaps and writes two clamped runtime descriptors after the dead-unit pad slot, patching `GLOBAL_DIM ord=2` to `cu_k[B]` device-side — exactly what `build_thd_meta_o_kv_descs_kernel` has always done for FP8/MXFP8. All four f16 SM100 flavors (d128, d192_d128, d256, d512) bind `tma_k`/`tma_v` to closures over those descriptors under `CFG.THD_VARLEN`. Tile-tail rows past the packed total become TMA-OOB and land as exact zeros: no memory read, no fill kernel, nothing written into the caller's buffer.

Measured on B200 / SM100, cuDNN 9.26.0.33, `CUDNN_FRONTEND_ENABLE_FROST_ENGINES=1` — same box, same libs, same command, only the clamp differing:

| suite | without clamp | with clamp |
|---|---|---|
| `test_mhas_v2 -k fwd_ragged_L0` | 19 failed / 27 passed | **0 failed / 46 passed** |
| `test_mhas_v2 -k fwd` (dense + ragged + fp8 + mxfp8) | — | **212 passed**, 28 skipped |
| FROST SM100 fwd sweep | — | **722 passed**, 0 failed |
| new `test_dsl_sm100_thd_nan_capacity_tail` | 201,728 / 403,456 / 806,912 NaNs in O at d128/d256/d512 | clean |

**Still open on this issue:**

- **SM120.** `prefill_f16_sm120.py` uses the meta-only `build_thd_meta_kernel`, so the K/V descriptor slots must be *introduced* rather than switched over. I had no SM120 hardware to verify against, so it is deliberately out of scope for #646 rather than shipped blind.
- **The follow-on noted earlier** — extending the same patch to Q from `cu_q[B]` would make Q/K/V extents all exact device-side (O is already per-sequence exact, LSE predicated per row), at which point `_thd_capacity` degrades from "must be ≥ the total **and** ≤ what the caller owns" to just "≤ what the caller owns", which is the only thing buffer geometry can honestly promise. Q does not need it for NaN-safety — it is the parallel dimension, so a garbage Q row poisons only its own row and that row is never stored — but it would remove the last host-capacity dependency. Related: #737 (padded THD would escalate this from one tail to B tails).
